README.md
<!--
    This file is part of PBala (http://github.com/oscarsaleta/PBala)

    Copyright (C) 2016  O. Saleta

    PBala is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->


# PBala (Princess Bala): DISTRIBUTED EXECUTION SOFTWARE FOR ANTZ

# Current version

Current version is 6.0.3. 

# Download and install
Go to the [latest release](https://www.github.com/oscarsaleta/PBala/releases/latest "Latest release") for download links.

To install PBala, simply extract the contents of one of the attached downloads in the release page and copy the executables `PBala` and `PBala_task` to the directory where you want to work (you can ignore the `include` directory).

Also, make sure you have these lines in your `.bashrc` (for _sh_ users):
```
export PVM_ROOT=/usr/lib/pvm3
export PVM_DPATH=$PVM_ROOT/lib/pvmd
```
or this in your `.cshrc` (for _csh_ users):
```
setenv PVM_ROOT /usr/lib/pvm3
```

# Examples
See [the examples directory](Examples "Examples") the examples directory.

# Changelog
See [the changelog](CHANGELOG.md "CHANGELOG.md").

# Explanation of the software

## Introduction

This C program uses PVM libraries in order to create a parallellization interface for

* **Maple** scripts
* **C** programs
* **Python** scripts
* **Pari/GP** scripts (although Pari should be also supported through the Pari C/C++ library or through gp2c by manually editing the output C code and compiling as a C program using the command provided by gp2c. The executable can then be run using the C module of this software.)
* **Sage** scripts
* **Octave** scripts

This interface lets the user execute a same script/program over multiple input data in several CPUs located at the antz computing server. It sports memory management so nodes do not run out of RAM due to too many processes being started in the same node. It also reports resource usage data after execution.

## Manual compilation and installation

### Dependencies
* [Parallel Virtual Machine (PVM)](http://www.netlib.org/pvm3/ "PVM homepage").
* C compiler
* CMake (version 2.6 or higher)

### Instructions

* Clone this git repository: `git clone https://github.com/oscarsaleta/PBala.git` in a directory of your system,
* Change dir into *PBala*: `cd PBala`,
* Create a build directory: `mkdir build && cd build`,
* Run the CMake command: `cmake ..`,
* Compile the code with `make`, the executables *PBala* and *PBala_task* will be found in *build/src*.
* Optionally, perform a system wide install: `sudo make install`, which will install *PBala* and *PBala_task* into */usr/local/bin* and *PBala_config.h* into */usr/local/include* (this requires root permissions).


## Documentation

Run `doxigen` and take a look at *html/index.html* for documentation.

## Usage

The program admits standard `--help` (`-?`), `--usage` and `--version` (`-V`) arguments.

Output from `./PBala --help`:

```
Usage: PBala [OPTION...] programflag programfile datafile nodefile outdir
PBala -- PVM SPMD execution parallellizer.
	programflag argument can be: 0 (Maple), 1 (C), 2 (Python), 3 (Pari), 4 (Sage),
or 5 (Octave)

  -c, --custom-process=/path/to/exec
                             Specify a custom path for the executable program
  -e, --create-errfiles      Create stderr files
  -g, --create-memfiles      Create memory files
  -h, --create-slavefile     Create node file
  -k, --kill                 Kill remainig PBala/PVM processes (WARNING: use at
                             own risk! Use only if something goes wrong during
                             an execution and PVM stops working and you have no
                             other important processes running)
  -m, --max-mem-size=MAX_MEM Max memory size of a task (KB)
  -s, --maple-single-core    Force single core Maple
  -?, --help                 Give this help list
      --usage                Give a short usage message
  -V, --version              Print program version

Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.

Report bugs to <osr@mat.uab.cat>.
```

Output from `./PBala --usage`:

```
Usage: PBala [-ceghs?V] [-m MAX_MEM] [--create-errfiles] [--create-memfiles]
            [--create-slavefile] [--max-mem-size=MAX_MEM] [--maple-single-core]
            [--help] [--usage] [--version]
            programflag programfile datafile nodefile outdir
```

Mandatory arguments explained:

* `programflag`:
    + 0 = Maple
    + 1 = C
    + 2 = Python
    + 3 = Pari
    + 4 = Sage
    + 5 = Octave
* `programfile`: path to program file
* `datafile`: path to data file
    + Line format is "tasknumber,arg1,arg2,...,argN"
* `nodefile`: path to PVM node file
    + Line format is "nodename number_of_processes"
* `outdir`: path to output directory

Options explained:

- `-c, --custom-process=/path/to/exec`: Specify a custom path for the Maple/Python/etc executable to use
- `-e, --create-errfiles`: Save stderr output for each execution in a task_stderr.txt file
- `-g, --create-memfiles`: Save memory info for each execution in a task_mem.txt file
- `-h, --create-slavefile`: Save a log of which task is given to which slave in node_info.txt
- `-m, --max-mem-size=MAX_MEM`: max amount of RAM (in KB) that a single execution can require
    + Remark: the default behaviour is not giving work to a slave unless more than 15% of the max memory is available
- `-s, --maple-single-core`: Force Maple to use a single core for its executions



### Conventions

 - Computer Algebra Systems (CAS): **Maple**, **PARI** and **Sage**:
    + For Maple, we define 2 variables: `taskId` and `taskArgs`. taskId is an identifier for the task number that we are sending to the Maple script. `taskArgs` are the actual arguments that Maple has to use to do the computations. It is important to use these names because they are passed to Maple this way.
    + PARI and Sage are executed by creating auxiliary scripts where `taskId` and `taskArgs` are defined, so they could be directly used in the scripts just like in Maple.
- Programming languages: **C**, **Python**:
    + For C and Python we use the `argv` arrays so make sure the program can read and use those variables (and perform the error checking because this software has no way of knowing if the data file is suitable for your program).
- Octave:
    + We define the `taskArgs` list as a _cell array_, so you need to use the proper indexing and accessors to use the stored values (see [the Examples section](Examples/README.md#octave "Octave section in Examples") for more information).

### Procedure of execution

There is no need of starting PVM using a hostfile (it is actually advised not to do so). Instead, execute the program and we will start PVM for you, using the information extracted from the nodefile.

## Memory reporting
  
- *Maple*: Maple uses multithreading to parallelize the executions by default. This is good for performance but bad for resource management, because the PVM task loses control of the processes spawned. Therefore, the output files task_mem\*.txt don't show accurate values for resource usage of the program, because we can only track the parent Maple process, which doesn't do any work besides spawning and controlling its child processes. 
    + *Workaround*: we execute Maple with the `-t` flag, so in \*\_err.txt (output error file) we can see the `kernelopts` line that reports memory usage and computation time directly from Maple.
- *C*, *Python*, *Pari* and *Sage*: as long as the program to be executed is sequential, the PVM task will be able to get resource usage from the execution (using C system call `getrusage()`) and print it to the task_mem\*.txt file.

## Error codes

If *PBala* aborts during execution you can check the return value in the command line to know which kind of error happened. You can also check the output file for more information.

Existing codes:

* **PBala.c**
    - 10: error with command line arguments passed to the program
    - 11: error when counting number of lines of nodefile
    - 12: error when trying to open nodefile
    - 13: error when trying to read nodefile
    - 14: error when getting the current working directory
    - 15: error when asking pvmd for my TID (task identifier)
    - 16: error when asking pvmd for the task parent
    - 17: error when counting number of lines of datafile
    - 18: error when creating output file
    - 19: error when spawning pvm tasks
    - 20: error when reading first column of datafile (must be task id)
    - 21: error when creating out_dir/node_info.txt (maybe out_dir does not exist?)
    - 22: invalid task number
    - 23: error when writing to a file (PARI/GP or Sage aux script)
    - 24: error when sanitizing Maple script for single CPU execution
    - 25: error when there is a duplicate PVM host and PBala fails to remove it
* **PBala_task.c**
    - 10: error when forking process
    - 11: task killed by system


CHANGELOG.md
<!--
    This file is part of PBala (http://github.com/oscarsaleta/PBala)

    Copyright (C) 2016  O. Saleta

    PBala is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->

# Releases and changelog
* **v6.0.4**:
    - Added new node `a09` to `PBala -k` command.
* **v6.0.3**:
    - Solved bug that caused PBala output to not be correctly redirected to a file in case of an execution with redirection (`PBala ... > file`).
* **v6.0.2**:
    - Solved bug that caused PBala to crash after executions, leaving PVM in a potentially corrupt state for following executions.
* **v6.0.1**:
    - Solved bug that caused the `PBala` master process to use 100% CPU while waiting for slaves to finish their work.
* **v6.0.0**:
    - Total overhaul of how PBala handles tasks. Before it used to be file I/O based: get one line, pass it as arguments to a PVM slave, etc. Now we have a linked list of tasks, we only read the input datafile once, and we are able to dynamically remove the completed tasks from the list or readd the tasks that were unable to complete due to error/lack of memory.
    - PBala can look for the executable `PBala_task` in the current directory and in ~/bin/. If it does not find the executable in these directories, it will print a clear error.
    - Redesign of logging functions, so PBala output looks cleaner and easier to understand.
* **v5.1.0**:
    - Add new option to PBala: `-c, --custom-process=/path/to/exec` for specifying a custom path to the Maple/Sage/Python/etc executable to use
    - Changed timers to nanoseconds resolution
* **v5.0.0**:
    - Added support for Octave executions (`task_id = 5`).
* **v4.2.0**:
    - Added `-k` option. If selected, PBala kills every process called `PBala`, `PBala_task`, `pvmd`, `python`, `mserver`, `sage` and `bash` from every node in antz. Also, it removes `/tmp/pvm*` from every node in antz. (This only affects the current user unless it is run as root, which should never be done).
    - If PVM fails to start due to `PvmDupHost` (duplicate host, likely because of a previous failed execution), PBala tries to remove the host socket and start the daemon again. It will do so twice, and if it fails, it will return with an error (`E_PVM_DUP=25`).
* **v4.1.3**:
    - Solved an issue that prevented the user from select Maple executions.
* **v4.1.2**:
    - Changed build system to *CMake* (build instructions in [readme](README.md "README")).
    - Added error check for Maple single CPU file preparation.
    - Added parentheses in two conditional expressions to avoid compilation warnings.
    - Removed unused variable.
* **v4.1.1**:
    - Added error checks in some unchecked `fgets`, `sscanf` and `fopen` calls.
    - Some source files renamed from `antz_*` to `PBala_*`.
* **v4.1.0**:
    - Added *autotools* support (`configure && make && make install`).
    - Solved issue [#1](https://github.com/oscarsaleta/PBala/issues/1 "Issue 1").
* **v4.0.0**:
    - Revamped argument parsing (see the whole new *Usage* section).
    - Changed default safety memory threshold from 25% of total RAM to 15% of total RAM (user caution advised).
    - Added flags for creating \*_stderr.txt files, \*_mem.txt files, and node_info.txt file. Default behaviour now is *not* to create them unless specified with the optional flags.
* **v3.0.2**:
    - Bugfix: if the program exited with error status after PVM was initialized, it did not halt the virtual machine. Now it does automatically, so no manual halt is needed.
* **v3.0.1**:
    - Changed procedure for cleaning auxiliary Pari/Sage scripts generated during execution.
* **v3.0.0**:
    - Added Sage support using `task_type=4`. Sage scripts must have extension *.sage*, *.py* or *.spyx* in order for Sage to execute them.
* **v2.0.1**:
    - Solved bug with executions that had less tasks than nodes (execution would not finish unless killed)
* **v2.0.0**:
 * **Automatic generation of unfinished tasks datafile** (i.e. a clone of datafile containing only the lines that correspond to tasks that were killed before they could finish execution). This means that the program can later be run again using this new data file as input instead of having to manually search for the necessary data.
    - Added more error checks and error cases when forking tasks.
* **v1.3.0**:
    - Added more info at the beginning of stderr (system call, executable, datafile, nodefile, output directory)
    - Added a version variable in antz_lib.h
* **v1.2.1**:
    - Solved a long standing segfault problem that triggered when the output_dir did not exist.
    - Added one more error code
* **v1.2.0**:
    - Added error codes (see bottom of this file) to better inform of what made your execution crash.
* **v1.1.0**:
    - Added ability to detect if a task was killed or stopped.
    - Changed format files, now there are 4 different files: task\*_stdout, task\*_stderr, task\*_mem and task\*_killed (this last one just for tasks that *have been killed*). This way all the files for the same task can be easily located.
    - Fixed a bug that caused nodefiles of a single node to stop the execution. 
* **v1.0.0**: First release
    - New name: *Princess Bala* (because this software makes Antz work much harder).
    - Paths no longer need to be absolute.
    - PVM path is the current dir, so the executables work out of the box.

antz_errcodes.h
/*! \file errcodes.h
 * \brief Header for error codes in PBala
 * \author Oscar Saleta Reig
 */

/* ERROR CODES FOR PBala.c MAIN RETURN VALUES */
#ifndef E_ARGS
#define E_ARGS 10
#endif

#ifndef E_NODE_LINES
#define E_NODE_LINES 11
#endif

#ifndef E_NODE_OPEN
#define E_NODE_OPEN 12
#endif

#ifndef E_NODE_READ
#define E_NODE_READ 13
#endif

#ifndef E_CWD
#define E_CWD 14
#endif

#ifndef E_PVM_MYTID
#define E_PVM_MYTID 15
#endif

#ifndef E_PVM_PARENT
#define E_PVM_PARENT 16
#endif

#ifndef E_DATAFILE_LINES
#define E_DATAFILE_LINES 17
#endif

#ifndef E_OUTFILE_OPEN
#define E_OUTFILE_OPEN 18
#endif

#ifndef E_PVM_SPAWN
#define E_PVM_SPAWN 19
#endif

#ifndef E_DATAFILE_FIRSTCOL
#define E_DATAFILE_FIRSTCOL 20
#endif

#ifndef E_OUTDIR
#define E_OUTDIR 21
#endif

#ifndef E_WRONG_TASK
#define E_WRONG_TASK 22
#endif


/* ERROR CODES FOR task.c SLAVE RETURN STATUS */
#ifndef ST_FORK_ERR
#define ST_FORK_ERR 10
#endif

#ifndef ST_TASK_KILLED
#define ST_TASK_KILLED 11
#endif
